pkg/util/containers/collectors/detector.go

```diff
@@ -0,0 +1,145 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package collectors
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+// ErrNothingYet is returned when no collector is currently detected.
+// This might change in the future if new collectors are valid.
+var ErrNothingYet = &retry.Error{
+	LogicError:    errors.New("No collector detected"),
+	RessourceName: "detector",
+	RetryStatus:   retry.FailWillRetry,
+}
+
+// ErrPermaFail is returned when all candidates have permanently failed.
+var ErrPermaFail = &retry.Error{
+	LogicError:    errors.New("No collector available"),
+	RessourceName: "detector",
+	RetryStatus:   retry.PermaFail,
+}
+
+// Detector holds the logic to initialise collectors, with retries,
+// and selecting the most appropriate among available collectors
+type Detector struct {
+	candidates         map[string]Collector
+	detected           map[string]Collector
+	preferredCollector Collector
+	preferredName      string
+}
+
+// NewDetector returns a Detector ready to use. If configuredName
+// is empty, autodetection is enabled. If not, only the one name
+// will be tried.
+func NewDetector(configuredName string) *Detector {
+	d := &Detector{
+		candidates: make(map[string]Collector),
+		detected:   make(map[string]Collector),
+	}
+	// Load candidates from catalog
+	for n, f := range defaultCatalog {
+		if configuredName != "" && n != configuredName {
+			// If a collector name is configured, skip the others
+			continue
+		}
+		d.candidates[n] = f()
+	}
+	return d
+}
+
+// GetPreferred detects, ranks and returns the best collector for now.
+// Result might change if new collectors are valid after start, then
```

Comment:
How would the process agent react to a "live" collector change, would it be seemless in the live container view?

Reply:
- Best case scenario: `container_id` is constant, the transition should be seamless (rates are computed by the process-agent code, not the backend, that provides monotonic counts)

- Worst-case scenario: `container_id` churns (would be a bug): previous (now duplicates) containers will expire from the live containers view in a few runs, as if they were stopped

Note: once all candidates have been retried (5 minutes), the collector is stable